include/swift/Runtime/Metadata.h

```diff
@@ -1760,7 +1758,15 @@ struct TargetFunctionTypeMetadata : public TargetMetadata<Runtime> {
   TargetPointer<Runtime, const Argument> getArguments() const {
     return reinterpret_cast<TargetPointer<Runtime, const Argument>>(this + 1);
```

Comment:
@rjmccall I'm working on the changes to separate parameters from flags when stored in the entry of the cache but I need some help understanding how to do that, especially what `this + 1` really means here? I've read comment in `SimpleGlobalCache` which talks about how `getExtraAllocationSize` is used, so here is what I think I need to do to make it work, please correct me if I'm wrong:

1. Modify `getExtraAllocationSize` so it tail-allocates enough space for array of `uint32_t` equal to number of parameters like this

```
static size_t getExtraAllocationSize(Key key) {
    auto numParams = key.getFlags().getNumArguments();
    return key.getFlags().getNumArguments() * (sizeof(FunctionTypeMetadata::Argument) + sizeof(uint32_t));
  }
```
2. Add a method to `FunctionCacheEntry`:

```
  TargetPointer<Runtime, uint32_t *> getParameterFlags() {
    return reinterpret_cast<TargetPointer<Runtime, uint32_t *>>(__address here__);
  }
```

Should __address here__ be `this + (numParams * sizeof(TargetFunctionTypeMetadata::Argument)) + 1` ?

Reply:
Remember that pointer arithmetic in C is scaled by the size of the pointee type.  ``this + 1`` is an idiom used when you allocate extra memory after an object ("extra" meaning "not just ``sizeof(T)``"), because it very conveniently creates a pointer to that extra memory without needing to write out the type of ``this``.  You then generally need to cast that pointer to some other type in order to do further pointer arithmetic.

So address probably needs to be ``reinterpret_cast<char*>(this + 1) + (numParams * sizeof(TargetFunctionTypeMetadata::Argument))``.

Reply:
Or, alternatively, ``reinterpret_cast<Argument*>(this + 1) + numParams``.

Reply:
Thank you!